tests/optional.cpp

```diff
@@ -392,6 +475,43 @@ TEST_CASE("optional::and_then()", "[optional]")
     // move semantics work
     i = std::move(move_only).and_then([&](auto value_ptr) { return string_to_digit(*value_ptr); });
     i.if_set([](int value) { REQUIRE(value == 3); }).if_unset([]() { FAIL("and_then must return an engaged optional when the optional is engaged and the functor succeeds"); });
+
+    SECTION("argument forwarding")
+    {
+        // converts a string into a number from 0-9, or an empty optional if the
+        // string wasn't a number from 0-9 then adds the second int value to the string value.
+        auto sum_string_and_digit = [](const std::string& str, const int& value) -> optional<int> {
+            if (str.size() == 1 && str[0] >= '0' && str[0] <= '9') {
+                return (str[0] - '0') + value;
+            } else {
+                return std::nullopt;
+            }
+        };
+
+        optional<std::string> s;
+        int test_argument = 2;
+        // and_then on an empty optional will always return an empty optional
+        // of the functor's result's type
+        optional<int> i = s.and_then([&](auto value, auto& test_argument) { return sum_string_and_digit(value, test_argument); }, test_argument);
+        REQUIRE(i.empty());
+
+        // empty optional if the functor returns nullopt
+        s = "not a number";
+        i = s.and_then([&](auto value, auto& test_argument) { return sum_string_and_digit(value, test_argument); }, test_argument);
+        REQUIRE(i.empty());
+
+        // "5" will succeed in string_to_digit, so the conversion must succeed
+        s = "5";
+        i = s.and_then([&](auto value, auto& test_argument) { return sum_string_and_digit(value, test_argument); }, test_argument);
+        i.if_set([](int value) { REQUIRE(value == 7); }).if_unset([]() { FAIL("and_then must return an engaged optional when the optional is engaged and the functor succeeds"); });
+
+        optional move_only { std::make_unique<std::string>("3") };
+
+        // move semantics work
+        i = std::move(move_only).and_then([&](auto value_ptr, auto& test_argument) { return sum_string_and_digit(*value_ptr, test_argument); }, test_argument);
+        i.if_set([](int value) { REQUIRE(value == 5); }).if_unset([]() { FAIL("and_then must return an engaged optional when the optional is engaged and the functor succeeds"); });
```

Comment:
same, prefer `REQUIRE(i == 5)`

Reply:
For these I noticed there are other places in the code that I had not changed that I could also update to match this. Do you still want those cases to throw errors and just only have the "section(argument forwarding)" to use this format? Or should all unit tests just use this format?

Reply:
you're free to change them anywhere you see that pattern, but I just meant for the new code you've introduced